nipype/pipeline/plugins/multiproc.py

```diff
@@ -134,207 +118,174 @@ def __init__(self, plugin_args=None):
 
-        # Check to see if a job is available
-        currently_running_jobids = np.flatnonzero((self.proc_pending == True) & \
-                                (self.depidx.sum(axis=0) == 0).__array__())
+        # Check to see if a job is available (jobs without dependencies not run)
+        # See https://github.com/nipy/nipype/pull/2200#discussion_r141605722
+        jobids = np.nonzero(~self.proc_done & (self.depidx.sum(0) == 0))[1]
+        jobids = np.flatnonzero(
+            ~self.proc_done & (self.depidx.sum(axis=0) == 0).__array__())
 
         # Check available system resources by summing all threads and memory used
-        busy_memory_gb = 0
-        busy_processors = 0
-        for jobid in currently_running_jobids:
-            if self.procs[jobid]._interface.estimated_memory_gb <= self.memory_gb and \
-                            self.procs[jobid]._interface.num_threads <= self.processors:
+        free_memory_gb, free_processors = self._check_resources(self.pending_tasks)
 
-                busy_memory_gb += self.procs[jobid]._interface.estimated_memory_gb
-                busy_processors += self.procs[jobid]._interface.num_threads
+        logger.info('Currently running %d tasks, and %d jobs ready. '
+                    'Free memory (GB): %0.2f/%0.2f, Free processors: %d/%d',
+                    len(self.pending_tasks), len(jobids),
+                    free_memory_gb, self.memory_gb, free_processors, self.processors)
 
-            else:
-                raise ValueError(
-                    "Resources required by jobid {0} ({3}GB, {4} threads) exceed what is "
-                    "available on the system ({1}GB, {2} threads)".format(
-                        jobid, self.memory_gb, self.processors,
-                        self.procs[jobid]._interface.estimated_memory_gb,
-                        self.procs[jobid]._interface.num_threads))
-
-        free_memory_gb = self.memory_gb - busy_memory_gb
-        free_processors = self.processors - busy_processors
-
-        # Check all jobs without dependency not run
-        jobids = np.flatnonzero((self.proc_done == False) & \
-                                (self.depidx.sum(axis=0) == 0).__array__())
-
-        # Sort jobs ready to run first by memory and then by number of threads
-        # The most resource consuming jobs run first
-        jobids = sorted(jobids,
-                        key=lambda item: (self.procs[item]._interface.estimated_memory_gb,
-                                          self.procs[item]._interface.num_threads))
-
-        if str2bool(config.get('execution', 'profile_runtime')):
-            logger.debug('Free memory (GB): %d, Free processors: %d',
-                         free_memory_gb, free_processors)
-
-        # While have enough memory and processors for first job
-        # Submit first job on the list
+        if len(jobids) + len(self.pending_tasks) == 0:
+            logger.debug('No tasks are being run, and no jobs can '
```

Comment:
shouldn't this raise an Exception?

Reply:
I was still trying to understand why the execution goes through this, but it actually happens (see https://654-17021985-gh.circle-artifacts.com/2/home/ubuntu/work/logs/example_fmri_spm_nested_MultiProc__data_examples__l2pipeline/pypeline.log and https://654-17021985-gh.circle-artifacts.com/2/home/ubuntu/work/logs/example_fmri_spm_nested_MultiProc__data_examples__level1/pypeline.log).

So this happens, and it seems to recover itself without getting trapped. I think we could keep the debug trace (meaning, do not elevate it to warning) and add a counter. If the counter reaches a threshold, then raise.

WDYT?

Reply:
seems to me that `self.proc_pending` and `self.pending_task` are not in sync.

Reply:
in that case, if no jobs are available in line 197 should we return similar to your no resources are available? after all this function's goal is to send jobs to workers.

Reply:
Will do. I'm not seeing this in any of the runs, only few cases in the very last loop.